assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,563 @@
+use crate::MODULE_PATH_DELIM;
+
+use super::{AssemblyError, Token, TokenStream, Vec};
+
+use std::collections::BTreeMap;
+
+mod types;
+use types::{Instruction, Node, Program};
+
+mod io_ops;
+mod stack_ops;
+mod u32_ops;
+
+#[cfg(test)]
+pub mod tests;
+
+// AST Parser that holds internal state to generate correct ASTs.
+#[allow(dead_code)]
+struct Parser {
+    uses: BTreeMap<String, String>,
+}
+
+impl Parser {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new instance of [Parser] that parses Miden programs into a AST.
+    #[allow(dead_code)] 
+    pub fn new() -> Self {
+        Self {
+            uses: BTreeMap::new(),
+        }
+    }
+
+    // PROGRAM COMPILER
+    // --------------------------------------------------------------------------------------------
+
+    // Compiles the provided source into a AST Program. This Program holds information
+    // that can be directly translated into MAST.
+    #[allow(dead_code)] 
+    pub fn compile(&mut self, source: &str) -> Result<Program, AssemblyError> {
+        let mut tokens = TokenStream::new(source)?;
+
+        let mut program = Program {
+            body: Vec::<Node>::new(),
+            procedures: BTreeMap::new(),
+        };
+
+        self.parse_imports(&mut tokens)?;
+        self.parse_procedures(&mut tokens, &mut program)?;
+
+        // make sure program body is present
+        let next_token = tokens
+            .read()
+            .ok_or_else(|| AssemblyError::unexpected_eof(tokens.pos()))?;
+        if next_token.parts()[0] != Token::BEGIN {
+            return Err(AssemblyError::unexpected_token(next_token, Token::BEGIN));
+        }
+
+        // parse the program body
+        self.parse_body(&mut tokens, &mut program)?;
+
+        // make sure there are no instructions after the end
+        if let Some(token) = tokens.read() {
+            return Err(AssemblyError::dangling_ops_after_program(token));
+        }
+
+        Ok(program)
+    }
+
+    // Parses the token streams into AST nodes
+    fn parse_imports(&mut self, tokens: &mut TokenStream) -> Result<(), AssemblyError> {
+        // read tokens from the token stream until all `use` tokens are consumed
+        while let Some(token) = tokens.read() {
+            match token.parts()[0] {
+                Token::USE => {
+                    let module_path = &token.parse_use()?;
+                    let (_, short_name) = module_path.rsplit_once(MODULE_PATH_DELIM).unwrap();
+                    self.uses
+                        .insert(short_name.to_string(), module_path.to_string());
```

Comment:
There should be check here to prevent declaration of duplicate modules.

Reply:
fixed